bfsscp/bfsscp.go

```diff
@@ -0,0 +1,376 @@
+//   )
+//
+//   func main() {
+//     ctx := context.Background()
+//     b, _ := bfs.Connect(ctx, "ssh://user:pass@hostname:22/path/to/root?tmpdir=%2Fcustom%2Ftmp")
+//     f, _ := b.Open(ctx, "file/within/root.txt")
+//     ...
+//   }
+//
+// bfs.Connect supports the following query parameters:
+//
+//   tmpdir - custom temp dir
+//
+package bfsscp
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/url"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/bmatcuk/doublestar"
+	"github.com/bsm/bfs"
+	"github.com/bsm/bfs/internal"
+	"github.com/kr/fs"
+	"github.com/pkg/sftp"
+	"go.uber.org/multierr"
+	"golang.org/x/crypto/ssh"
+)
+
+// Register allows for registering more schemes with SCP
+func Register(_ context.Context, u *url.URL) (bfs.Bucket, error) {
+	query := u.Query()
+	address := net.JoinHostPort(u.Host, u.Port())
+
+	username, password := "", ""
+	if u.User != nil {
+		username = u.User.Username()
+		password, _ = u.User.Password()
+	}
+
+	return New(address, &Config{
+		Username: username,
+		Password: password,
+		Prefix:   u.Path,
+		TempDir:  query.Get("tmpdir"),
+	})
+}
+
+func init() {
+	bfs.Register("ssh", Register)
+	bfs.Register("scp", Register)
+}
+
+// Config is passed to New to configure the SSH connection.
+type Config struct {
+	// Username to use.
+	Username string
+	// Password to use in combination with Username.
+	Password string
+	// An optional path prefix.
+	Prefix string
+	// A custom temp dir.
+	TempDir string
+}
+
+func (c *Config) norm() error {
+	c.Prefix = strings.TrimPrefix(c.Prefix, "/")
+	if c.Prefix != "" && !strings.HasSuffix(c.Prefix, "/") {
+		c.Prefix = c.Prefix + "/"
+	}
+	return nil
+}
+
+type bucket struct {
+	conn   *ssh.Client
+	client *sftp.Client
+	config *Config
+}
+
+// New initiates an bfs.Bucket backed by ssh.
+func New(address string, cfg *Config) (bfs.Bucket, error) {
+	config := new(Config)
+	if cfg != nil {
+		*config = *cfg
+	}
+	if err := config.norm(); err != nil {
+		return nil, err
+	}
+
+	sshConfig := &ssh.ClientConfig{
+		User: cfg.Username,
+		Auth: []ssh.AuthMethod{
+			ssh.Password(cfg.Password),
+		},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+
+	conn, err := ssh.Dial("tcp", address, sshConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := sftp.NewClient(conn)
+	if err != nil {
+		return nil, multierr.Combine(conn.Close(), err)
+	}
+
+	return &bucket{
+		conn:   conn,
+		client: client,
+		config: config,
+	}, nil
+}
+
+func (b *bucket) withPrefix(name string) string {
+	if b.config.Prefix == "" {
+		return name
+	}
+	return internal.WithinNamespace(b.config.Prefix, name)
+}
+
+// Glob implements bfs.Bucket.
+func (b *bucket) Glob(_ context.Context, pattern string) (bfs.Iterator, error) {
+	// quick sanity check
+	if _, err := doublestar.Match(pattern, ""); err != nil {
+		return nil, err
+	}
+
+	// create a new iterator
+	return &iterator{
+		w:       b.client.Walk(b.withPrefix("/")),
```

Comment:
1. you should pass the context to the iterator
2. this is obviously not very efficient and we only need to Walk the full file tree if we have a `**` in the pattern somewhere. I am not too fuzzed about it at the moment but if you have an idea, let me know

Reply:
Btw, glob-ing files from `xxx://bucket/path/with*pattern` is not very convenient across all the adapters (you have to manually instantiate a bucket and then glob files there), would be nice to solve it for all of them.
IIRC, I've seen some code that just iterates entire bucket contents on each Glob-ing without any clever prefix-ing.

Reply:
@dim any path with `**` now uses the `walkerIterator` all others use sftp's internal `Glob` and `matchesIterator`